File: NativeChart/android/src/main/java/org/github/henryquan/nativechart/line/NativeLineChart.kt
package org.github.henryquan.nativechart.line

import android.graphics.Color
import com.facebook.react.uimanager.ThemedReactContext
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.LimitLine
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet


class NativeLineChart(c: ThemedReactContext) : LineChart(c) {
  var chartData: ArrayList<Float> = arrayListOf()
  var legendLabel: String = ""
  var darkMode: Boolean = false
  var themeColor: Int = Color.WHITE

  init {
    this.setupChart()
  }

  /**
   * Update chart
   */
  fun updateChart() {
    // Remove all lines
    this.axisRight.removeAllLimitLines()

    // Map float to entry
    val formattedData =
      this.chartData.mapIndexed { index, element -> Entry(index.toFloat(), element) }

    // Setup data and colour to chart
    val dataSetList = mutableListOf<ILineDataSet>()
    val chartDataSet = LineDataSet(formattedData, this.legendLabel)
    chartDataSet.color = this.themeColor
    chartDataSet.setCircleColor(this.themeColor)
    chartDataSet.circleRadius = 3.0F
    chartDataSet.setDrawValues(false)
    dataSetList.add(chartDataSet)

    val chartDataFormatted = LineData(dataSetList)
    this.data = chartDataFormatted

    // Update theme colour
    val textColour = if (darkMode) Color.WHITE else Color.BLACK
    this.axisLeft.textColor = textColour
    this.legend.textColor = textColour

    // Add average line
    val avg = chartData.reduce { acc, curr -> acc + curr } / chartData.size.toFloat()
    val average = LimitLine(avg, String.format("%.1f", avg))
    average.labelPosition = LimitLine.LimitLabelPosition.RIGHT_BOTTOM
    average.lineWidth = 0.5F
    average.textColor = textColour
    // Fit to theme colour
    average.lineColor = this.themeColor
    this.axisRight.addLimitLine(average)
  }

  /**
   * Setup the chart
   */
  private fun setupChart() {
    // Text related
    this.setNoDataText("...")
    this.description.text = ""

    // Disable zoom and interaction
    this.isHighlightPerTapEnabled = false
    this.isHighlightPerDragEnabled = false
    this.setScaleEnabled(false)

    // Custom style for xAxia
    this.xAxis.position = XAxis.XAxisPosition.BOTTOM
    this.xAxis.setDrawGridLines(false)
    this.xAxis.setDrawLabels(false)

    // Custom style for axisRight
    this.axisLeft.setDrawGridLines(false)

    // Custom style for axisLeft
    this.axisRight.setDrawGridLines(false)
    this.axisRight.setDrawLabels(false)
    this.axisRight.setDrawAxisLine(false)
  }
}

File: NativeChart/ios/Charts/Line/NativeLineChart.swift
//
//  NativeLineChart.swift
//  NativeChart
//
//  Created by Yiheng Quan on 12/2/20.
//

import Charts

class NativeLineChart : LineChartView, OptimisedBarLineChart {
    
    @objc var chartData: [Double] = []
    @objc var legendLabel: String = ""
    @objc var darkMode: Bool = false
    @objc var themeColor: String = "#FFFFFF"
    
    override func didSetProps(_ changedProps: [String]!) {
        super.didSetProps(changedProps)
        self.updateChart()
    }
    
    init() {
        super.init(frame: CGRect())
        setupChart(chart: self)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    /** Update chart data whenever props are passed from js */
    internal func updateChart() {
        // Remove all lines to prevent drawing multiple lines
        self.rightAxis.removeAllLimitLines()

        // Map double to ChartDataEntry
        let formattedData = self.chartData.enumerated().map({ i, element in ChartDataEntry(x: Double(i), y: element)})
        
        // Setup data and colour to chart
        let realThemeColour = UIColor(hex: themeColor)!
        let chartDataSet = LineChartDataSet(entries: formattedData, label: legendLabel)
        chartDataSet.setColor(realThemeColour)
        chartDataSet.setCircleColor(realThemeColour)
        chartDataSet.circleRadius = 3.0
        chartDataSet.drawValuesEnabled = false
        let chartDataFormatted = LineChartData(dataSets: [chartDataSet])
        self.data = chartDataFormatted
        
        // Update theme colour
        let textColour = darkMode ? UIColor.white : UIColor.black
        self.leftAxis.labelTextColor = textColour
        self.legend.textColor = textColour
        
        // Add average line
        let avg = chartData.reduce(0.0, { x, y in x + y }) / Double(chartData.count)
        let average = ChartLimitLine(limit: avg, label: String(format: "%.1f", avg))
        average.labelPosition = .rightBottom
        average.lineWidth = 0.5
        average.valueTextColor = textColour
        // Fit to theme colour
        average.lineColor = realThemeColour
        self.rightAxis.addLimitLine(average)
    }

    /** Some basic styling for LineChartView */
    internal func setupChart(chart: BarLineChartViewBase) {
        // Text related
        chart.noDataText = "No Information are provided"
        chart.chartDescription.text = ""
        
        // Disable zoom and interaction
        chart.highlightPerTapEnabled = false
        chart.highlightPerDragEnabled = false
        chart.setScaleEnabled(false)

        // Custom style for xAxia
        chart.xAxis.labelPosition = .bottom
        chart.xAxis.drawGridLinesEnabled = false
        chart.xAxis.drawLabelsEnabled = false
        
        // Custom style for rightAxis
        chart.rightAxis.drawLabelsEnabled = false
        chart.rightAxis.drawGridLinesEnabled = false
        chart.rightAxis.drawAxisLineEnabled = false
        
        // Custom style for leftAxis
        chart.leftAxis.drawLabelsEnabled = true
        chart.leftAxis.drawGridLinesEnabled = false
        chart.leftAxis.drawAxisLineEnabled = true
    }
}
